Some improvements to the hydroflow template

rohitkulshreshtha opened the pull request:
1. Default server address, so you don't have to copy an address around when starting multiple clients to test a full star network configuration.
2. Help messages for all the arguments.
3. More hand-holding documentation, mostly condensed from the [first networking chapter](https://hydro.run/docs/hydroflow/quickstart/example_7_echo_server)
4. Updated README.md to provide more runnable commands.

rohitkulshreshtha committed:
Some improvements to the hydroflow template

1. Default server address, so you don't have to copy an address around when starting multiple clients to test a full star network configuration.
2. Help messages for all the arguments.
3. More hand-holding documentation, mostly condensed from the [first networking chapter](https://hydro.run/docs/hydroflow/quickstart/example_7_echo_server)
4. Updated README.md to provide more runnable commands.

rohitkulshreshtha commented on README.md:
````diff
@@ -1,86 +1,47 @@
+cargo run -- --help
 ```
 
-The `src/main.rs` file is where the command-line arguments are parsed and the appropriate service is launched.
-It also contains the `Opts` struct, which uses the [clap](https://docs.rs/clap/latest/clap/) crate to 
-specify the command-line arguments that are accepted by the application.
-It is possible to change the command-line arguments by modifying the `Opts` struct.
-Before launching the service, the `main` function binds to the specified address and prints out the address that was bound to.
-
-The `src/protocol` file contains the enum `Message`, which can include messages with very different structures. 
-The `Message` must provide the `Serialize` and `Deserialize` traits, which are used by the [serde](https://docs.serde.rs/serde/) crate.
-In the application, the `Message` enum includes an `Echo` message that has a (`String`) payload and timestamp; it also includes
-`Heartbeat` and `HeartbeatAck` messages that carry no information other than their type. (Messages are delivered with the sender 
-address attached, so these empty message types can be useful.) 
-
-The `src/helpers.rs` file contains any helper functions that are needed. In our example this is just a function to print
-a dataflow graph representation of the hydroflow code to stdout.
-
-Each service file comes with a skeleton Hydroflow spec that provides an inbound communication channel and an outbound communication channel, both 
-bound to the specified address and port. The channels are named `inbound_chan` and `outbound_chan`, and are accessed in hydroflow code using the 
-`source_serde` and `sink_serde` operators respectively. The single address/port pair is sufficient in general to support multiple
-different `Message` types across multiple services and instances. Upon receipt, messages are handled by the appropriate code using hydroflow's `demux`
-operator. It is also possible to open more channels to segregate traffic to different IP addresses or ports, simply by copying the 
-patterns that define and use the `inbound` and `outbound` channels.
````
I have inlined most of this within the code itself. I am hoping it becomes easier to read it at the same site as the code itself.

rohitkulshreshtha commented on README.md:
```diff
@@ -1,86 +1,47 @@
 
-Each service file also includes code to generate the dataflow graph for the service, if the `--graph` flag is provided on the command line.
-The ASCII spec for the graph is printed to stdout on launch.
+| File          | Description                                                                                                                          | 
+|---------------|--------------------------------------------------------------------------------------------------------------------------------------|
+| `main.rs`     | Contains `main` entry-point function for both client and server. Performs command-line argument parsing.                             |
+| `protocol.rs` | Contains the `Message` enum that defines the messages that can be sent between instances.                                            |
+| `<role>.rs`   | Contains the service for the given role. Example implementations and skeletal hydroflow spec are provided for `server` and `client`. |
+| `helpers.rs`  | Contains helper functions that are invoked from Hydroflow code in multiple services.                                                 |
```
I hope y'all are OK with tables. RustRover markdown editor makes tables really easy, but if y'all feel strongly about not dealing with tables, I can revert this back to the original text blcok.

MingweiSamuel replied:
Tables are 👌

rohitkulshreshtha commented on README.md:
```diff
@@ -100,3 +61,31 @@ In our experience, when starting a Hydroflow project we recommend a four-step ap
 
 Have fun!
+
+## Print a Dataflow Graph
```
I moved this section to the bottom because it seemed like an appropriate order/priority of information to me.

rohitkulshreshtha commented on src/server.rs:
```diff
@@ -1,50 +1,83 @@
-            ->  demux(|(msg, addr), var_args!(echo, heartbeat, errs)|
-                    match msg {
+        inbound_demuxed = inbound_chan
+            -> inspect(|(m, a): &(Message, SocketAddr)| println!("{}: Got {:?} from {:?}", Utc::now(), m, a)) // For debugging purposes.
```
Thought this was a good place to introduce `inspect`. It removes the `tee` from the program, but this is OK because an [earlier chapter](https://hydro.run/docs/hydroflow/quickstart/example_6_unreachability) introduces it.

rohitkulshreshtha commented:
My guess is that `cargo fmt --check` was added recently? For example, it is failing on 

```rust
Echo { payload: String, ts: DateTime<Utc> }
```

which has always existed...

MingweiSamuel commented on src/main.rs:
```diff
@@ -10,51 +10,66 @@ mod helpers;
-    Dot,
-}
-
+/// A simple echo server & client generated using the Hydroflow template. The lines starting with
+/// `///` contain the message that appears when you run the compiled binary with the '--help'
+/// arguments, so feel free to change it to whatever makes sense for your application.
```
Can also link to the `clap` docs 
https://docs.rs/clap/latest/clap/

MingweiSamuel commented on src/main.rs:
```diff
@@ -10,51 +10,66 @@ mod helpers;
-    #[clap(long, value_parser = ipv4_resolve)]
-    addr: Option<SocketAddr>,
-    // #[clap(long)]
-    #[clap(long, value_parser = ipv4_resolve)]
-    server_addr: Option<SocketAddr>,
-    #[clap(value_enum, long)]
-    graph: Option<GraphType>,
+    graph: Option<GraphType>, // This too, is an optional argument.
```
Actually, would be good to update this to use the args from `use hydroflow::lang::graph::{WriteConfig, WriteGraphType};`

```rust
    #[clap(long)]
    graph: Option<WriteGraphType>,
    #[clap(flatten)]
    write_config: Option<WriteConfig>,
```

Example usage:

https://github.com/hydro-project/hydroflow/blob/550b17c0e0edd233b0cceab55fed0de309051503/hydroflow/examples/kvs/server.rs#L36-L41

rohitkulshreshtha replied:
Done!

MingweiSamuel commented on src/main.rs:
```diff
@@ -10,51 +10,66 @@ mod helpers;
+
+/// The default server address & port on which the server listens for incoming messages. Clients
+/// send message to this address & port.
+pub const DEFAULT_SERVER_ADDRESS: &str = "localhost:54321";
```
Wonder if it would be worth using `cargo template` to generate a random port each generation to avoid collisions

rohitkulshreshtha replied:
Since the example is for running a single server, think this should be OK. If you feel strongly about this, I'd prefer to take it up separately.

MingweiSamuel commented:
Oh and for cargo fmt, just run `cargo fmt` (or `cargo +nightly fmt`) and it will fix all the formatting and pass CI

rohitkulshreshtha committed:
Addressed feedback.

jhellerstein reviewed (commented):
LGTM. Maybe change the earlier advice to use `for_each(...println!...)` to say `inspect(...println!...)`?

rohitkulshreshtha committed:
Get clippy passing.

rohitkulshreshtha committed:
Update ci.yml

The hydroflow template testing step needs to be updated to use the address simplification.
